---
title: "Cell type identification"
output: github_document
---

Created by: Philip Lijnzaad

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=9, fig.height=6) 
```

# Overview

In any single-cell experiment where mixed populations are expected, a
very important task is to identify the cell types (and/or cell states)
that are present. In this practical, we will look at two tools for doing
this: SingleR ([Aran et
al. 2019](https://www.nature.com/articles/s41590-018-0276-y)) and CHETAH
([de Kanter et al., under review](http://dx.doi.org/10.1101/558908)).

# Datasets

We will try to classify the ovarian ascites data from [Schelker et
al. 2017](https://www.nature.com/articles/s41467-017-02289-3) using, as
a reference, the different reference data sets from both SingleR and CHETAH.


The `ovarian1200` dataset contains 1200 cells randomly selected from the
3114 single cell data provided by Schelker et al. (which is available
from https://figshare.com/s/711d3fb2bd3288c8483). The data come from the
ascites of 4 different ovarian cancer patients and contain a variety of
leukocytes as well as tumor cells.


First load the various packages:

```{r packages}
suppressMessages(require(SingleCellExperiment))
suppressMessages(require(Seurat))
suppressMessages(require(CHETAH))
suppressMessages(require(Matrix))
suppressMessages(require(SingleR))
suppressMessages(require(cowplot))
```

If you have (or downloaded) the ovarian data into folder
`data.dir` then load the `Seurat` object with

```{r load_ovarian}
#load expression matrix

data.dir <- "session-celltypeid_files" # or wherever the data is located
file <- paste0(data.dir,"/ovarian1200.rds")
ovarian <- readRDS(file=file)

```

The authors already classified the cells 'by hand' using marker genes. This classfication can be found in the `celltypes` column of the `meta.data` of the object. Get an overview of this:

```{r overian_overview}

head(ovarian@meta.data)

sort(table(ovarian@meta.data$celltypes))

```

Clearly, most of the cells are Macrophages. 

We now follow the 'standard' [Seurat workflow](https://satijalab.org/seurat/essential_commands.html#seurat-standard-worflow) to prepare the data. 

```{r prepare_ovarian, warning=FALSE}
ovarian <- NormalizeData(object = ovarian)
ovarian <- FindVariableFeatures(object = ovarian)
ovarian <- ScaleData(object = ovarian)
ovarian <- RunPCA(object = ovarian, npcs=20)
ovarian <- FindNeighbors(object = ovarian)
ovarian <- FindClusters(object = ovarian, resolution=0.5) 
## (the default resolution yields too manh clusters)
ovarian <- RunTSNE(object = ovarian)
ovarian <- RunUMAP(object = ovarian, dims=1:20)
p1 <- DimPlot(ovarian, reduction='tsne')
p2 <- DimPlot(ovarian, reduction='umap')
plot_grid(p1, p2)
```

This shows the data with Seurat clustering, but we're more interested in the cell types. Let's see if and how they coincide. For convenience and flexibility, we define `dim.red` to be our dimension reduction of choice.

```{r cluster_vs_publ_types}
dim.red <- 'tsne' # or 'umap' or 'pca'
p1 <- DimPlot(ovarian, group.by="seurat_clusters", reduction=dim.red)
p2 <- DimPlot(ovarian, group.by='celltypes', reduction=dim.red)
plot_grid(p1, p2, labels=c('clusters', 'published types'))
```

# SingleR

## SingleR reference data sets

We will now do our cell type identification using SingleR. SingleR
comes with a collection of reference data sets. There are two for human:
`hpca`, the Human Primary Cell Atlas (microarray-based), and
`blueprint\_encode`, a combined Blueprint Epigenomics and Encode data
set (RNASeq based) .

For mouse there are also two sets: `immgen`, the Immunological Genome
Project (microarray-based) and `mouse.rnaseq`, a brain specific reference
(RNASeq-based). For details I refer to the SingleR-specifications
vignette.

Each reference data set comes in two 'flavours': 'label.fine', which are very
fine-grained, detailed types, and 'label.main', which are a coarser
subset of those types.

The SingleR reference data sets are part of the package, and can be
explored easily.

```{r explore_singler_refs}

warning("now using celldex!")

library(celldex)

hpca <- HumanPrimaryCellAtlasData()
table(hpca$label.main)

table(hpca$label.fine)

bpe <- BlueprintEncodeData()
table(bpe$label.main)
table(bpe$label.fine)

```

## Using SingleR with other reference data sets

SingleR only needs a single gene expression profile per cell type, which
makes it possible to use bulk-RNAsequencing and even micorarrays as
reference data. The downside is that the variability within cell types
is not represented (although their methods do provide a p-value. Again, see 
SingleR's highly recommend vignettes).

The other method we will look at, CHETAH, needs several (100-200)
single-cell expression profiles for the classification. The advantage is
that the inherent variability is fully account for. CHETAH, originally
developed for working with cancer data has its 'own' reference data set

that is based on single-cell data from Head-Neck cancer, melanoma,
breast and colorectal cancer. (For details see
https://figshare.com/s/aaf026376912366f81b6) Note that it is easy to
create your own reference data sets for both SingleR and CHETAH.

SingleR can use the CHETAH reference if that has been 'bulkified' by
averaging over all cells per reference cell type. We provide this as a
ready-made object (`chetah.ref.singler`).

The layout of the reference data is quite simple: a `list` with the name
of the reference, a big data matrix (genes x celltypes), and the types
per cell, both in a detailed version (`$label.fine`) and the simple version
(`$label.main`) For the CHETAH reference we duplicated the (`$types`)
and the simple version (`$main_types`). (Note that the
`chetah.ref.singler` reference object can only be used by SingleR, not
by CHETAH).

```{r chetah_ref_singler}

file <- paste0(data.dir, "/chetah.ref.singler.rds")
chetah.ref.singler <- readRDS(file=file)

# which main type are there:
unique(chetah.ref.singler$main_types)

# layout of the object:
str(chetah.ref.singler)

```

## Classifying with SingleR

SingleR can classify using several different reference data sets at the
same time; this saves time and memory.

On to the actual classification with SingleR.

```{r singler_classification}

hpca <- HumanPrimaryCellAtlasData()

counts <- GetAssayData(ovarian, assay="RNA", slot="data")

singler <- SingleR(test=counts, ref=hpca, labels=hpca$label.main)

```

The `ref.list` argument specified a named list with three different
reference data sets: HPCA, blueprint\_encode ('bpe') and the bulkified
chetah\_reference ('snglr\_chetah'). The resulting `singler` object has
complete classifications for each of these reference sets, under the
`$singler` member. The actual types per cell are found in sub-list
`$SingleR.single.main$labels[,1]`. 

(side-note: SingleR also automatically classifies per cluster of
cells, but we will not use this type of classification.)

To get a good overview it's easiest to iterate over all elements of this
list.

```{r explore_snglr_results}

show(table(singler$first.labels))

## For hpca and blueprint_encode also show the 
## detailed cell typings (as opposed to main_types results) : 



### Session info

```{r sessioninfo}
sessionInfo()
```
